lib/ui/text/asset_manager_font_provider.cc

```diff
@@ -0,0 +1,119 @@
+#include "third_party/skia/include/core/SkString.h"
+#include "third_party/skia/include/core/SkTypeface.h"
+
+namespace blink {
+
+namespace {
+
+void VectorReleaseProc(const void* ptr, void* context) {
+  delete reinterpret_cast<std::vector<uint8_t>*>(context);
+}
+
+}  // anonymous namespace
+
+AssetManagerFontProvider::AssetManagerFontProvider(
+    fml::RefPtr<blink::AssetManager> asset_manager)
+    : asset_manager_(asset_manager) {}
+
+AssetManagerFontProvider::~AssetManagerFontProvider() = default;
+
+size_t AssetManagerFontProvider::GetFamilyCount() {
+  return family_names_.size();
+}
+
+std::string AssetManagerFontProvider::GetFamilyName(int index) {
+  return family_names_[index];
+}
+
+SkFontStyleSet* AssetManagerFontProvider::MatchFamily(
+    const std::string& family_name) {
+  auto found = registered_families_.find(family_name);
+  if (found == registered_families_.end()) {
+    return nullptr;
+  }
+  return SkRef(&found->second);
+}
+
+void AssetManagerFontProvider::RegisterAsset(std::string family_name,
+                                             std::string asset) {
+  auto family_it = registered_families_.find(family_name);
+
+  if (family_it == registered_families_.end()) {
+    family_names_.push_back(family_name);
+    family_it = registered_families_
+                    .emplace(std::piecewise_construct,
+                             std::forward_as_tuple(family_name),
+                             std::forward_as_tuple(asset_manager_))
+                    .first;
+  }
+
+  family_it->second.registerAsset(asset);
+}
+
+AssetManagerFontStyleSet::AssetManagerFontStyleSet(
+    fml::RefPtr<blink::AssetManager> asset_manager)
+    : asset_manager_(asset_manager) {}
+
+AssetManagerFontStyleSet::~AssetManagerFontStyleSet() = default;
+
+void AssetManagerFontStyleSet::registerAsset(std::string asset) {
+  assets_.emplace_back(asset);
+}
+
+int AssetManagerFontStyleSet::count() {
+  return assets_.size();
+}
+
+void AssetManagerFontStyleSet::getStyle(int index, SkFontStyle*,
+                                        SkString* style) {
+  FXL_DCHECK(false);
+}
+
+SkTypeface* AssetManagerFontStyleSet::createTypeface(int i) {
+  size_t index = static_cast<size_t>(i);
+  if (index >= assets_.size())
+    return nullptr;
+
+  TypefaceAsset& asset = assets_[i];
+  if (!asset.typeface) {
+    std::unique_ptr<std::vector<uint8_t>> asset_buf =
+        std::make_unique<std::vector<uint8_t>>();
+    if (!asset_manager_->GetAsBuffer(asset.asset, asset_buf.get()))
+      return nullptr;
+
+    std::vector<uint8_t>* asset_buf_ptr = asset_buf.release();
+    sk_sp<SkData> asset_data = SkData::MakeWithProc(
+        asset_buf_ptr->data(), asset_buf_ptr->size(), VectorReleaseProc,
+        asset_buf_ptr);
+    std::unique_ptr<SkMemoryStream> stream = SkMemoryStream::Make(asset_data);
+
+    // Ownership of the stream is transferred.
+    asset.typeface = SkTypeface::MakeFromStream(stream.release());
```

Comment:
This ctor takes an index, are you sure your don't have to pass the index onto the typeface?

TBH, I am not entirely clear how this mechanism with the indices works.

Reply:
The index just identifies the font within this font set, ranging from 0 through count() - 1

Reply:
Gotcha